deliver/lib/deliver/upload_screenshots.rb

```diff
@@ -9,17 +9,44 @@ def upload(options, screenshots)
       v = app.edit_version
       UI.user_error!("Could not find a version to edit for app '#{app.name}'") unless v
 
-      UI.message("Starting with the upload of screenshots...")
+      # md5's of uploaded screenshots
+      @checksums_local = {}
+      # md5's of ITC screenshots
+      @checksums_remote = {}
+
+      screenshots.each do |screenshot|
+        # checksum of uploaded screenshot
+        md5 = Spaceship::Utilities.md5digest(screenshot.path)
+        @checksums_local[screenshot.language] ||= {}
+        # to keep order we have to split them per device
+        device_type = screenshot.device_type
+        @checksums_local[screenshot.language][device_type] ||= []
+        @checksums_local[screenshot.language][device_type] << md5
+      end
 
-      # First, clear all previously uploaded screenshots, but only where we have new ones
-      # screenshots.each do |screenshot|
-      #   to_remove = v.screenshots[screenshot.language].find_all do |current|
-      #     current.device_type == screenshot.device_type
-      #   end
-      #   to_remove.each { |t| t.reset! }
-      # end
+      v.screenshots.each do |lang, screenshots_for_lang|
+        screenshots_for_lang.each do |current|
+          matched = current.original_file_name.match(/ftl_[0-9a-f]{32}_*/)
+          checksum = current.original_file_name.split('_').at(1) if matched
+
+          # store remote checksum. We will need it later to determine if we have to upload screenshot
+          @checksums_remote[current.language] ||= {}
+          @checksums_remote[current.language][current.device_type] ||= []
+          @checksums_remote[current.language][current.device_type] << checksum
+
+          # Remove from ITC non existing locally screenshots
+          # Or screenshots that have wrong order (compare indexes)
+          # Or screenshots that we haven't uploaded through spaceship
+          unless checksum && local_and_remote_match_to_delete(current, checksum)
+            UI.message("Deleting screenshot #{current.original_file_name} for language #{current.language}")
```

Comment:
not sure if original_file_name helps (especially if it has the ftl_..._ helps).

Maybe do something like 
```
original_file_name = current.original_file_name
matched = original_file_name.match(/ftl_([0-9a-f]{32})_(*)/)
if matched 
  checksum = matched[1]
  original_file_name = matched[2]
end
...

  UI.message("Deleting screenshot ##{current.sort_order} #{original_file_name} ...."
```